### PR TITLE
Feature/95 thing rebase

### PR DESCRIPTION
**Pull request by benjamineac**

### CODE REVIEW

File: graphql_api/schema/custom/automation_task_base.py [hunk 1]
```diff
@@ -58,20 +58,6 @@ class AutomationTaskInterface(graphene.Interface):
...
-    def resolve_parents(self, info, **args):
-        print('AutomationTaskBaseParentResolver')
-        print(self.parents)
-        t0 = dt.utcnow()
-        if not self.parents:
-            res = []
-        elif (len(info.field_asts[0].selection_set.selections)==1 and
-            info.field_asts[0].selection_set.selections[0].name.value == 'total_count'):
-            from graphql_api.schema.task_task_relation import TaskTaskRelationConnection
-            return TaskTaskRelationConnection(edges=[None for x in range(len(self.parents))])
```
Comment (benjamineac) [inline graphql_api/schema/custom/automation_task_base.py]: removed unused resolver

File: graphql_api/schema/custom/general_task.py [hunk 2]
```diff
@@ -66,40 +66,6 @@ def resolve_swept_arguments(self, info, **args):
...
                     yield itm['k']
 
-    def resolve_children(self, info, **args):
-        t0 = dt.utcnow()
-        if not self.children:
-            res = []
-        elif (len(info.field_asts[0].selection_set.selections)==1 and
-            info.field_asts[0].selection_set.selections[0].name.value == 'total_count'):
-            from graphql_api.schema.task_task_relation import TaskTaskRelationConnection
-            res = TaskTaskRelationConnection(edges= [None for x in range(len(self.children))])
```
Comment (benjamineac) [inline graphql_api/schema/custom/general_task.py]: same here

File: graphql_api/schema/custom/rupture_generation_task.py [hunk 3]
```diff
@@ -37,17 +37,6 @@ class Meta:
     @staticmethod
     def from_json(jsondata):
         return RuptureGenerationTask(**AutomationTaskBase.from_json(jsondata))
-    
-    @staticmethod                            
-    def resolve_parents(root, info, *args, **kwargs):
-        if not root.parents:
-            res = []
-        elif isinstance(root.parents[0], dict):
```
Comment (benjamineac) [inline graphql_api/schema/custom/rupture_generation_task.py]: and here

File: graphql_api/schema/thing.py [hunk 4]
```diff
@@ -54,6 +54,10 @@ def resolve_parents(root, info, **args):
         t0 = dt.utcnow()
         if not root.parents:
             res = []
+        elif (len(info.field_asts[0].selection_set.selections)==1 and
+            info.field_asts[0].selection_set.selections[0].name.value == 'total_count'):
+            from graphql_api.schema.task_task_relation import TaskTaskRelationConnection
+            res = TaskTaskRelationConnection(edges= [None for x in range(len(root.parents))])
         elif isinstance(root.parents[0], dict):
             #new form, parents is list of objects
             res = [get_data_manager().thing.get_one(parent['parent_id']) for parent in root.parents]
```
Comment (benjamineac) [inline graphql_api/schema/thing.py:63]: added total count optimisation to thing resolvers

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -38,7 +38,7 @@ pip install -r test_requirements.txt
 
 and validate our tests are passing:
 ```
-SLS_OFFLINE=1 pytest
+SLS_OFFLINE=1 TESTING=1 DEPLOYMENT_STAGE=TEST pytest
```
Comment (chrisbc) [inline DEVELOPMENT.md:41]: @benjamineac  maybe this file is redundant now? It's mostly duplicating out new README

File: graphql_api/config.py [hunk 6]
```diff
@@ -21,3 +24,4 @@ def boolean_env(environ_name, default='FALSE'):
 DEPLOYMENT_STAGE = os.getenv('DEPLOYMENT_STAGE', 'LOCAL').upper()
 STACK_NAME = os.getenv('STACK_NAME')
 CW_METRICS_RESOLUTION = os.getenv('CW_METRICS_RESOLUTION', 60) #1 for high resolution or 60
+S3_BUCKET_NAME = os.environ.get('S3_BUCKET_NAME', "nshm-tosh-api-test")
```
Comment (chrisbc) [inline graphql_api/config.py:27]: @benjamineac  - safer to make the default "S3_BUCKET_NAME_unconfigured".

File: graphql_api/schema/custom/automation_task_base.py [hunk 7]
```diff
@@ -45,6 +33,7 @@ def get_node(cls, info, _id):
     @staticmethod
     def from_json(jsondata):
         #Field type transforms...
+        started = dt.now()
```
Comment (chrisbc) [inline graphql_api/schema/custom/automation_task_base.py:36]: @benjamineac I don't think from_json should be setting started  ??

File: graphql_api/schema/task_task_relation.py [hunk 8]
```diff
@@ -18,14 +18,23 @@ class Meta:
...
+    # class Meta:
+    #     interfaces = (relay.Node, )
 
+    parent = graphene.Field(GeneralTask, required=False)
+    child = graphene.Field(ChildTaskUnion, required=False)
+    
     parent_id = graphene.String()
     child_id = graphene.String()
+    
+    @staticmethod
```
Comment (chrisbc) [inline graphql_api/schema/task_task_relation.py:30]: @benjamineac are these two resolvers actually needed as they only return root.

File: graphql_api/schema/thing.py [hunk 9]
```diff
@@ -44,6 +49,46 @@ def resolve_files(root, info, **args):
             db_metrics.put_duration(__name__, 'resolve_files[legacy]' , dt.utcnow()-t0)
 
         return res
+    
+    def resolve_parents(root, info, **args):
```
Comment (chrisbc) [inline graphql_api/schema/thing.py:53]: @benjamineac my concern about having this here is that we're saying that the parent<->child relationship is represented as a TaskTaskRelationConnection. 

IMO Thing is super primitive, with no built in relationships except the files. Whereas parents & children are attributes of a Task (which should be a subclass of Thing).

File: graphql_api/schema/custom/automation_task_base.py (hunk 7 above)
Comment (benjamineac) [reply graphql_api/schema/custom/automation_task_base.py:36]: I didn't touch this function.. this looks weird to me

File: graphql_api/tests/test_s3_fallback.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        assert thing._read_object('0')['clazz_name'] == 'RuptureGenerationTask'
+       
+      
+    def test_thing_read_s3(self):
+        with mock_s3():
+            conn = boto3.resource('s3', region_name='us-east-1')
+            conn.create_bucket(Bucket=S3_BUCKET_NAME)
+            bucket = conn.Bucket(S3_BUCKET_NAME)
+            thing = ThingData(thing_args, self._data_manager, ToshiThingObject, self._connection)
+            thing.create(clazz_name='RuptureGenerationTask', created=dt.datetime.now(tzutc()))
```
Comment (chrisbc) [inline graphql_api/tests/test_s3_fallback.py:62]: I don't think this is needed?

**Review by chrisbc — CHANGES_REQUESTED** [review]
@benjamineac A few changes/questions to look at, plus:
 - we need to review serverless.yml. some stuff here was needed
 - need to look into parent/child on Task vs Thing.
 - create new object is not in a transaction AFAIK, but maybe I'm missing that in GH review

File: graphql_api/schema/custom/automation_task_base.py (hunk 7 above)
Comment (chrisbc) [reply graphql_api/schema/custom/automation_task_base.py:36]: weird indieed - scuttle it!

**Review by chrisbc — APPROVED** [review]

**Pull request merged**